File: src/services/chunkManager.ts
// Chunk Manager Service
// Split large documents into translation-friendly chunks

import type { Chunk } from '../types';

/**
 * Clause pattern for EN/ISO standards
 * Matches: 5, 5.1, 5.1.2, 5.1.2.3, A.1, B.2.3, etc.
 */
const CLAUSE_PATTERN = /^((?:\d+\.)*\d+|[A-Z]\.\d+(?:\.\d+)*)\s+/;

/**
 * Split text into chunks for translation
 * Uses clause-aware splitting for technical standards
 * Target: 500-1000 tokens per chunk
 */
export function splitIntoChunks(text: string, maxTokens: number = 800): Chunk[] {
    const chunks: Chunk[] = [];

    // Split into lines first
    const lines = text.split('\n');

    // Group lines into clause sections
    const sections: { clauseNumber: string | null; lines: string[] }[] = [];
    let currentSection: { clauseNumber: string | null; lines: string[] } = { clauseNumber: null, lines: [] };

    for (const line of lines) {
        const clauseMatch = line.match(CLAUSE_PATTERN);

        if (clauseMatch) {
            // New clause detected - save current section and start new one
            if (currentSection.lines.length > 0) {
                sections.push(currentSection);
            }
            currentSection = { clauseNumber: clauseMatch[1], lines: [line] };
        } else {
            // Continue current section
            currentSection.lines.push(line);
        }
    }

    // Don't forget the last section
    if (currentSection.lines.length > 0) {
        sections.push(currentSection);
    }

    // Now chunk each section
    let chunkPosition = 0;

    for (const section of sections) {
        const sectionText = section.lines.join('\n').trim();
        if (!sectionText) continue;

        const sectionTokens = estimateTokens(sectionText);

        if (sectionTokens <= maxTokens) {
            // Section fits in one chunk
            chunks.push(createChunk(sectionText, chunkPosition++, section.clauseNumber));
        } else {
            // Section too large - split by paragraphs within the section
            const paragraphs = sectionText.split(/\n\n+/).filter(p => p.trim());
            let currentChunkText = '';
            let isFirstParagraph = true;

            for (const paragraph of paragraphs) {
                const combinedTokens = estimateTokens(currentChunkText + '\n\n' + paragraph);

                if (currentChunkText && combinedTokens > maxTokens) {
                    // Save current chunk, start new one
                    chunks.push(createChunk(
                        currentChunkText,
                        chunkPosition++,
                        isFirstParagraph ? section.clauseNumber : null
                    ));
                    currentChunkText = paragraph;
                    isFirstParagraph = false;
                } else {
                    currentChunkText = currentChunkText
                        ? currentChunkText + '\n\n' + paragraph
                        : paragraph;
                }
            }

            // Add remaining text
            if (currentChunkText.trim()) {
                chunks.push(createChunk(
                    currentChunkText,
                    chunkPosition++,
                    isFirstParagraph ? section.clauseNumber : null
                ));
            }
        }
    }

    return chunks;
}

/**
 * Create a chunk object
 */
function createChunk(text: string, position: number, clauseNumber?: string | null): Chunk {
    const trimmedText = text.trim();
    const firstLine = trimmedText.split('\n')[0];

    // Detect chunk type using multiple signals
    const indicators = {
        // Markdown style Heading
        markdown: /^#{1,6}\s/.test(trimmedText),

        // Standard Numbered (1.2.3 Title)
        numbered: /^(\d+\.)+\d*\s+[A-Z]/.test(firstLine),

        // Capitalized Title Case (excluding single letters)
        capitalized: /^[A-Z][a-zA-Z\s]{2,}[^.!?]*$/.test(firstLine),

        // Short length (headings usually < 100 chars)
        short: firstLine.length < 100,

        // No sentence ending punctuation (unless it's a quote)
        noPunctuation: !/[.!?]$/.test(firstLine),

        // Explicit "Chapter"/"Section" start
        explicit: /^(Chapter|Section|Annex|Appendix)\s+\w+/i.test(firstLine)
    };

    // Calculate heading score
    let headingScore = 0;
    if (indicators.markdown) headingScore += 5; // Strongest signal
    if (indicators.explicit) headingScore += 5;
    if (indicators.numbered && indicators.short) headingScore += 3;
    if (indicators.capitalized && indicators.short && indicators.noPunctuation) headingScore += 2;

    // Clause number also indicates heading
    if (clauseNumber && indicators.short) headingScore += 3;

    // List detection
    const isList = /^[\-\*\u2022\d]+[\.\)]\s/.test(trimmedText) || trimmedText.startsWith('- ');

    // Table detection (Markdown style or many pipes)
    const tablePipeCount = (trimmedText.match(/\|/g) || []).length;
    const isTable = tablePipeCount > 4 && trimmedText.split('\n').filter(l => l.includes('|')).length > 2;

    let type: Chunk['type'] = 'paragraph';
    let metadata: Chunk['metadata'] = {};

    // Add clause number to metadata if present
    if (clauseNumber) {
        metadata.clauseNumber = clauseNumber;
    }

    if (headingScore >= 2) {
        type = 'heading';
        const headingMatch = trimmedText.match(/^(#{1,6})\s+(.+)/);
        if (headingMatch) {
            metadata.level = headingMatch[1].length;
            metadata.heading = headingMatch[2];
        } else {
            // Infer level from clause number depth
            metadata.heading = firstLine.replace(/^#{1,6}\s+/, ''); // Clean up
            if (clauseNumber) {
                metadata.level = clauseNumber.split('.').length;
            } else {
                metadata.level = indicators.markdown ? (trimmedText.match(/^#+/)?.[0].length || 2) : 2;
            }
        }
    } else if (isList) {
        type = 'list';
    } else if (isTable) {
        type = 'table';
    }

    return {
        id: `chunk_${position}`,
        text: trimmedText,
        position,
        type,
        metadata
    };
}

/**
 * Estimate token count (rough approximation)
 * ~1 token per 4 characters for English
 */
export function estimateTokens(text: string): number {
    return Math.ceil(text.length / 4);
}

/**
 * Reassemble translated chunks into full document
 */
export function reassembleChunks(chunks: { text: string; type: string }[]): string {
    return chunks
        .map((chunk) => {
            // Add spacing based on chunk type
            if (chunk.type === 'heading') {
                return `\n\n${chunk.text}\n`;
            }
            return chunk.text;
        })
        .join('\n\n')
        .trim();
}

/**
 * Calculate chunks statistics
 */
export function getChunkStats(chunks: Chunk[]) {
    const totalTokens = chunks.reduce((sum, chunk) => sum + estimateTokens(chunk.text), 0);
    const avgTokensPerChunk = Math.round(totalTokens / chunks.length);

    return {
        totalChunks: chunks.length,
        totalTokens,
        avgTokensPerChunk,
        types: {
            heading: chunks.filter(c => c.type === 'heading').length,
            paragraph: chunks.filter(c => c.type === 'paragraph').length,
            list: chunks.filter(c => c.type === 'list').length,
            table: chunks.filter(c => c.type === 'table').length
        }
    };
}

File: backend/services/chunk_manager.py
"""
Chunk Manager Service - Text chunking for translation.
"""

import re
from typing import Iterator
from models.requests import Chunk


def estimate_tokens(text: str) -> int:
    """Rough token estimation (1 token ≈ 4 characters for English)."""
    return len(text) // 4


def split_into_chunks(
    text: str,
    max_tokens: int = 1500,
    overlap_tokens: int = 100
) -> list[Chunk]:
    """
    Split document text into translation-friendly chunks.
    
    Respects:
    - Paragraph boundaries
    - Section headers
    - List blocks
    - Table blocks
    
    Args:
        text: Full document text
        max_tokens: Maximum tokens per chunk
        overlap_tokens: Token overlap between chunks for context
    
    Returns:
        List of Chunk objects
    """
    chunks = []
    
    # Split by double newlines (paragraphs) while preserving structure
    paragraphs = re.split(r'\n\n+', text)
    
    current_chunk = []
    current_tokens = 0
    chunk_index = 0
    
    for para in paragraphs:
        para = para.strip()
        if not para:
            continue
        
        para_tokens = estimate_tokens(para)
        
        # If single paragraph exceeds max, split it further
        if para_tokens > max_tokens:
            # Flush current chunk first
            if current_chunk:
                chunk_text = "\n\n".join(current_chunk)
                chunks.append(Chunk(
                    id=f"chunk_{chunk_index}",
                    content=chunk_text,
                    index=chunk_index
                ))
                chunk_index += 1
                current_chunk = []
                current_tokens = 0
            
            # Split large paragraph by sentences
            sentences = re.split(r'(?<=[.!?。！？])\s+', para)
            for sentence in sentences:
                sent_tokens = estimate_tokens(sentence)
                
                if current_tokens + sent_tokens > max_tokens:
                    if current_chunk:
                        chunk_text = " ".join(current_chunk)
                        chunks.append(Chunk(
                            id=f"chunk_{chunk_index}",
                            content=chunk_text,
                            index=chunk_index
                        ))
                        chunk_index += 1
                        current_chunk = []
                        current_tokens = 0
                
                current_chunk.append(sentence)
                current_tokens += sent_tokens
        
        # Normal case: paragraph fits
        elif current_tokens + para_tokens > max_tokens:
            # Flush current chunk
            if current_chunk:
                chunk_text = "\n\n".join(current_chunk)
                chunks.append(Chunk(
                    id=f"chunk_{chunk_index}",
                    content=chunk_text,
                    index=chunk_index
                ))
                chunk_index += 1
            
            # Start new chunk with this paragraph
            current_chunk = [para]
            current_tokens = para_tokens
        
        else:
            # Add to current chunk
            current_chunk.append(para)
            current_tokens += para_tokens
    
    # Don't forget the last chunk
    if current_chunk:
        chunk_text = "\n\n".join(current_chunk)
        chunks.append(Chunk(
            id=f"chunk_{chunk_index}",
            content=chunk_text,
            index=chunk_index
        ))
    
    return chunks


def merge_translated_chunks(chunks: list[dict]) -> str:
    """
    Merge translated chunks back into a single document.
    Handles overlap removal if needed.
    """
    sorted_chunks = sorted(chunks, key=lambda c: c.get("index", 0))
    
    texts = [chunk.get("translated", "") for chunk in sorted_chunks]
    
    # Simple join with double newlines
    return "\n\n".join(texts)
